lib/pmdtester/builders/rule_set_builder.rb

```diff
@@ -0,0 +1,107 @@
+# frozen_string_literal: true
+
+require 'nokogiri'
+require 'set'
+require_relative '../cmd'
+module PmdTester
+  # This class is responsible for generation dynamic configuration
+  # according to the difference between base and patch branch of Pmd.
+  class RuleSetBuilder
+    ALL_RULE_SETS = Set['bestpractices', 'codestyle', 'design', 'documentation', 'errorprone',
+                        'migrating', 'multithreading', 'performance', 'regex', 'security'].freeze
+    PATH_TO_PMD_JAVA_BASED_RULES =
+      'pmd-java/src/main/java/net/sourceforge/pmd/lang/java/rule/'
+    PATH_TO_PMD_XPATH_BASED_RULES = 'pmd-java/src/main/resources/category/java/'
+    PATH_TO_ALL_JAVA_RULES = 'config/all-java.xml'
+    PATH_TO_DYNAMIC_CONFIG = 'target/dynamic-config.xml'
+    NO_JAVA_RULES_CHANGED_MESSAGE = 'No java rules have been changed!'
+
+    def initialize(options)
+      @options = options
+    end
+
+    def build
+      filenames = diff_filenames
+      rule_sets = get_rule_sets(filenames)
+      build_config_file(rule_sets)
+    end
+
+    def filter_set(rule_sets)
+      if rule_sets.size.eql?(ALL_RULE_SETS.size)
+        # if `rule_sets` contains all rule sets, than no need to filter the baseline
+        @options.filter_set = nil
+      else
+        @options = rule_sets
+      end
+    end
+
+    def diff_filenames
+      filenames = nil
+      Dir.chdir(@options.local_git_repo) do
+        diff_cmd = "git diff --name-only #{@options.base_branch}..#{@options.patch_branch}"
```

Comment:
Do me a favor and add a little comment, that we only need to support git here, since PMD's repo is using git...
I caught myself thinking about the need for mercurial etc... :smile:

Reply:
Done :smile: